README
[![Build Status](https://travis-ci.org/tomac/yersinia.svg?branch=master)](https://travis-ci.org/tomac/yersinia)

Spanning Tree
-------------

#1: DOS attack sending conf BPDUs

    Let's send some conf BPDUs claiming be root!!! By sending continously conf BPDU with root pathcost 0, randomly
    generated bridge id (and therefore the same root id), and some default values for other fields, we try to 
    annoy the switches close to us, causing a DoS when trying to parse and recalculate their STP engines.

    Source MAC: randomly generated.
    Destination MAC: 01:80:c2:00:00:00
    Bridge ID: 8000:source_mac
    Root ID: 8000:source_mac
    Hello time: 2
    Forward delay: 15
    Max age: 20
    Root pathcost: 0

    <output from the cisco log>
    01:20:26: STP: VLAN0001 heard root 32768-d1bf.6d60.097b on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-9ac6.0f72.7118 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-85a3.3662.43dc on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-3d84.bc1c.918e on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-b2e2.1a12.dbb4 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-4ba6.2d45.5844 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-deb0.4f14.7288 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-4879.8036.0e24 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-2776.e340.9222 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-299e.de76.c07d on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-d38b.bc5b.e90d on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-78ee.0205.afdb on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-b32b.e969.81b1 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-b16b.c428.88a3 on Fa0/8
    01:20:26: STP: VLAN0001 heard root 32768-dd01.1436.9044 on Fa0/8
    </output>


#2: DOS attack sending tcn BPDUs

    This attack sends continously tcn BPDUs causing the root switch to send conf BPDUs acknowledging the change. 
    Besides, the root switch will send topology change notifications to the members of the tree, and they will
    have to recalculate their STP engine to learn the new change.

    Source MAC: randomly generated.
    Destination MAC: 01:80:c2:00:00:00

    <output from the cisco log>
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    01:35:39: STP: VLAN0001 Topology Change rcvd on Fa0/8
    </output>


#3: NONDOS attack Claiming Root Role

    Now our aim is to get the root role of the tree. How can we accomplish this issue? Just listening to the network
    to find out which one is the root role, and start sending conf BPDU with lower priority to become root.

    Source MAC: same one as the sniffed BPDU.
    Destination MAC: same one as the sniffed BPDU.
    Bridge ID: the sniffed one slightly modified to have a lower priority
    Root ID: 8000: same as bridge id.
    Hello time: same one as the sniffed BPDU.
    Forward delay: same one as the sniffed BPDU.
    Max age: same one as the sniffed BPDU.
    Root pathcost: same one as the sniffed BPDU.

    <output from the cisco log>
    01:58:48: STP: VLAN0001 heard root 32769-000e.84d4.2280 on Fa0/8
    01:58:48:     supersedes 32769-000e.84d5.2280
    01:58:48: STP: VLAN0001 new root is 32769, 000e.84d4.2280 on port Fa0/8, cost 19
    </output>


#4 NONDOS attack Claiming a non-root role

   We pretend to be another weird switch playing with STP and praising our root id :)


#5 DOS attack causing eternal root elections

    By sending config BPDUs autodecrementing their priority, we can cause infinite root elections in the STP tree.
    It would be something similar to recount the election's votes to determine the winner (do you remember Florida?)

    <output from the cisco log>
    00:20:21: STP: VLAN0001 heard root 32769-000e.84d4.2280 on Fa0/9
    00:20:21:     supersedes 32769-000e.84d5.2280
    00:20:21: STP: VLAN0001 new root is 32769, 000e.84d4.2280 on port Fa0/9, cost 19
    00:20:23: STP: VLAN0001 heard root 32769-000e.84d3.2280 on Fa0/9
    00:20:23:     supersedes 32769-000e.84d4.2280
    00:20:23: STP: VLAN0001 new root is 32769, 000e.84d3.2280 on port Fa0/9, cost 19
    00:20:25: STP: VLAN0001 heard root 32769-000e.84d2.2280 on Fa0/9
    00:20:25:     supersedes 32769-000e.84d3.2280
    00:20:25: STP: VLAN0001 new root is 32769, 000e.84d2.2280 on port Fa0/9, cost 19
    00:20:27: STP: VLAN0001 heard root 32769-000e.84d1.2280 on Fa0/9
    00:20:27:     supersedes 32769-000e.84d2.2280
    00:20:27: STP: VLAN0001 new root is 32769, 000e.84d1.2280 on port Fa0/9, cost 19
    00:20:29: STP: VLAN0001 heard root 32769-000e.84d0.2280 on Fa0/9
    00:20:29:     supersedes 32769-000e.84d1.2280
    00:20:29: STP: VLAN0001 new root is 32769, 000e.84d0.2280 on port Fa0/9, cost 19
    00:20:31: STP: VLAN0001 heard root 32769-000e.84cf.2280 on Fa0/9
    00:20:31:     supersedes 32769-000e.84d0.2280
    00:20:31: STP: VLAN0001 new root is 32769, 000e.84cf.2280 on port Fa0/9, cost 19
    00:20:33: STP: VLAN0001 heard root 32769-000e.84ce.2280 on Fa0/9
    00:20:33:     supersedes 32769-000e.84cf.2280
    00:20:33: STP: VLAN0001 new root is 32769, 000e.84ce.2280 on port Fa0/9, cost 19
    00:20:35: STP: VLAN0001 heard root 32769-000e.84cd.2280 on Fa0/9
    00:20:35:     supersedes 32769-000e.84ce.2280
    00:20:35: STP: VLAN0001 new root is 32769, 000e.84cd.2280 on port Fa0/9, cost 19
    00:20:37: STP: VLAN0001 heard root 32769-000e.84cc.2280 on Fa0/9
    00:20:37:     supersedes 32769-000e.84cd.2280
    00:20:37: STP: VLAN0001 new root is 32769, 000e.84cc.2280 on port Fa0/9, cost 19
    00:20:39: STP: VLAN0001 heard root 32769-000e.84cb.2280 on Fa0/9
    00:20:39:     supersedes 32769-000e.84cc.2280
    00:20:39: STP: VLAN0001 new root is 32769, 000e.84cb.2280 on port Fa0/9, cost 19
    </output>


#6 DOS Attack causing root dissapearance

    This time we try to exhaust the root election proccess. We manage to become root in the STP tree,
    but we stop sending config BPDUs until it reaches max_age seconds (usually 20), forcing a new 
    election proccess.

    <output from the cisco log>
    02:02:43: STP: VLAN0001 heard root 32769-000e.84d4.2280 on Fa0/9
    02:02:43:     supersedes 32769-000e.84d5.2280
    02:02:43: STP: VLAN0001 new root is 32769, 000e.84d4.2280 on port Fa0/9, cost 19
    02:03:03: STP: VLAN0001 we are the spanning tree root
    02:03:04: STP: VLAN0001 heard root 32769-000e.84d4.2280 on Fa0/9
    02:03:04:     supersedes 32769-000e.84d5.2280
    02:03:04: STP: VLAN0001 new root is 32769, 000e.84d4.2280 on port Fa0/9, cost 19
    02:03:04: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:06: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:08: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:10: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:12: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:14: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:16: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:18: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:20: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:22: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:24: STP: VLAN0001 we are the spanning tree root
    02:03:24: STP: VLAN0001 heard root 32769-000e.84d4.2280 on Fa0/9
    02:03:24:     supersedes 32769-000e.84d5.2280
    02:03:24: STP: VLAN0001 new root is 32769, 000e.84d4.2280 on port Fa0/9, cost 19
    02:03:24: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:26: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:28: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:30: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:32: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:34: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:36: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:38: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:40: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:42: STP: VLAN0001 sent Topology Change Notice on Fa0/9
    02:03:44: STP: VLAN0001 we are the spanning tree root
    </output>


Mitigations (Cisco only)
------------------------

- Use port security and disable STP in those ports that don't require STP.
For information about port security, please check the following url:
http://www.cisco.com/en/US/products/hw/switches/ps628/products_configuration_guide_chapter09186a0080150bcd.html

- If you are using the portfast feature in your STP configuration, enable also the BPDU guard for avoiding these
attacks when the port automatically enters the forwarding state:
http://www.cisco.com/warp/public/473/65.html

- Use the root guard feature for avoiding rogue devices to become root:
http://www.cisco.com/en/US/tech/tk389/tk621/technologies_tech_note09186a00800ae96b.shtml


Further reading
---------------

- Guillermo Marro's nice Master Thesis:
http://seclab.cs.ucdavis.edu/papers/Marro_masters_thesis.pdf

- Oleg K. Artemjev, Vladislav V. Myasnyankin. Fun with the Spanning Tree Protocol
http://phrack.org/issues/61/12.html



CDP 
---

Cisco devices have always spoken a different language to communicate among them,
to tell everybody that they are alive and which nifty features they have. CDP
stands for Cisco Discovery Protocol. By means of this particular language, Cisco
devices set up a virtual world where everyone is happy and ther is no crime at
all. Or at least it seems to be this way, since many network administrators
aren't worried for CDP yet. Although some of information that can be sent in a
CDP packet is still undocumented (CDP is a propietary protocol and it seems that
Cisco does not want to give details about it), at least one old attack (that it
is still valid) is known, and there is a public implementation (FX did it!).

This attack is a DoS trying to exhaust the device memory so that it can't
allocate more memory for any device process. How can we do it? Simply sending
CDP packets with bogus data simulating real Cisco devices. The target device
will begin to allocate memory in its CDP table to save the new neighbor
information, but without knowing that it is going to have thousands or millions
of new friends.

Other attack implemented in the current code is the ability to set up a new
virtual Cisco device that it is designated only for make a little mess, trying
to confuse network administrators.

Screenshot of the attack running:

Output of a Cisco 2503 router:

athens#sh mem
               Head   Total(b)    Used(b)    Free(b)  Lowest(b) Largest(b)
Processor     4873C    3893444    3893444          0          0          0
I/O          400000    2097152    2097088         64         64         64

<log>
%SCHED-3-THRASHING: Process thrashing on watched queue 'CDP packets' (count 57).
-Process= "CDP Protocol", ipl= 6, pid= 9
-Traceback= 3159232 31594DE 3201660
%LANCE-5-COLL: Unit 0, excessive collisions. TDR=7
%SCHED-3-THRASHING: Process thrashing on watched queue 'CDP packets' (count 57).
-Process= "CDP Protocol", ipl= 6, pid= 9
-Traceback= 3159232 31594DE 3201660
%SYS-2-MALLOCFAIL: Memory allocation of 100 bytes failed from 0x3201B3E, pool Processor, alignment 0
-Process= "CDP Protocol", ipl= 0, pid= 9
-Traceback= 314E8A4 314FA06 3201B46 32016C8
%SCHED-3-THRASHING: Process thrashing on watched queue 'CDP packets' (count 57).
-Process= "CDP Protocol", ipl= 6, pid= 9
-Traceback= 3159232 31594DE 3201660
%SYS-2-MALLOCFAIL: Memory allocation of 100 bytes failed from 0x3201B3E, pool Processor, alignment 0
-Process= "CDP Protocol", ipl= 0, pid= 9
-Traceback= 314E8A4 314FA06 3201B46 32016C8
%SYS-2-MALLOCFAIL: Memory allocation of 100 bytes failed from 0x3201B3E, pool Processor, alignment 0
-Process= "CDP Protocol", ipl= 0, pid= 9
-Traceback= 314E8A4 314FA06 3201B46 32016C8
</log>

And a couple of minutes later after killing the attack, the router surprisingly gets halted for several seconds, 
and then kicks you out of the terminal :)

<log>
%SYS-3-CPUHOG: Task ran for 16884 msec (69/69), Process = Exec, PC = 3158D42
-Traceback= 3158CEE 3158D4A 30F7330 30F742A 30FF3A4 30FEF76 30FEF1C 3116860
</log>


Output of a Cisco 2950 switch:

00:06:08: %SYS-2-MALLOCFAIL: Memory allocation of 224 bytes failed from 0x800118D0, alignment 0
Pool: Processor  Free: 0  Cause: Not enough free memory
Alternate Pool: I/O  Free: 32  Cause: Not enough free memory

-Process= "CDP Protocol", ipl= 0, pid= 26
-Traceback= 801DFC30 801E1DD8 800118D8 80011218 801D932C 801D9318
00:06:08:  ../src-calhoun/strata_stats.c at line 137: can't not push event list
00:06:09:  ../src-calhoun/strata_stats.c at line 137: can't not push event list

ChangeLog
CHANGES for Yersinia
====================

2017/08/24
----------
- Version 0.8.2
- FIX: CL: Fix reversed use of IP addresses parameters on command line, issue #44


2017/08/24
----------
- Version 0.8.1
- Fixed critical bug using invalid network interfaces (appeared initially with STP MiTM attack), issue #41


2017/08/23
----------
- Version 0.8.0
- Changed email addresses
- More code cleaning (renamed internal attack structures, removed cvs variable, ...)
- GTK: Reworked attack dialogs
- FIX: DHCP: Fixed "rogue server" attack, issue #12
- FIX: GTK: Memory leak using attack parameters, issue #38
- FIX: GTK: Attack parameters dialog was unable to close when launching an attack, issue #39
- FIX: GTK: Closing main window with the close button doesn't work, #36
- FIX: GTK: Out of bounds access refreshing main window, issue #35


2017/08/11
----------
- Version 0.8.0beta1
- Cleaned up some code (removed unused variables and macros, fixed compiler warnings, changed logging...)
- FIX: Error searching for pcap folder, thanks to digininja
- FIX: Changed pcap filter for sniffing of 802.1Q packets, issue #27
- FIX: CDP: Bad pointer usage, issue #14
- FIX: STP: Bad crafting of TCN packet, issue #32
- FIX: GTK: Bug on write_log that leads to random crashes, issue #9
- FIX: GTK: Usage of NULL variable using right click, issue #6
- FIX: GTK: Packet counters update, issue #26
- FIX: GTK: Add missing argument in function, issues #19, #20, #21, #22, #23 and #28
- FIX: GTK: Using nonexistant widget throws a critical error, issue #16
- FIX: GTK: Unable to close About window using the close button, issue #17
- FIX: GTK: Some toggled menues (MAC spoofing and others) don't work, issue #29
- FIX: GTK: Out of bound access on protocols array from buttons "Clear stats" and "Edit", issue #30
- FIX: GTK: Out of bound access on protocols array from button "Load default", issue #31
- FIX: GTK: Bad usage of attacks index, issues #24 and #25


2013/08/21
----------
- Version 0.7.3
- Fixed our pretty big issue with network interfacces under Linux. :)
- Fixed pcap_compile error.
- Refactoring a little bit of interfaces.c


2007/05/01
----------
- Version 0.7.2
- Fixed coredump on ncurses_i_ifaces_screen().
- Removed 'min' and 'max' members from all the structures.


2007/04/27
----------
- Added new protocol: MPLS. Rudimentary approach... :)


2007/02/08
----------
- Version 0.7.1
- Now it works in MacOSX 10.4 and perhaps in other BSD (BPF support)


2007/01/25
----------
- Added VTP exploit


2006/03/23
----------
- Forbidden launching attacks that need parameters from command line interface.
- Better 802.1X support.
- Better VTP printable data.
- Ncurses fixes


2006/03/22
----------
- We've got a new domain, www.yersinia.net :)
- Removed get_info protocol function.


2006/03/15
----------
- Removed update_data from protocols.


2006/03/14
----------
- Fixed ncurses show_info window if no protocol packets.
- Added alphabetical order for CLI. Yahooooo!!


2006/03/10
----------
- Added 802.1X simple MitM attack. :)


2006/03/09
----------
- Pre-pre-pre-pre-liminary release of 802.1X. :)
- Fixed 802.1Q problems with ncurses.


2006/02/17
----------
- Fixed configure bugs and better compilation for --disable-admin option. :)
- Updated man page (not finished yet).
- Fixed STP error when using RSTP.


2006/02/16
----------
- Fixed *big* race condition on Solaris due to the use of nonsafe
  multithreading calls. Yaaaahooo!!. :)


2006/02/09
----------
- Now it works again in OpenBSD (don't ignore your parents)
- Fixed lots of bugs compiling on Solaris platform.
- Added minimal protocol stats for the CLI.


2006/02/05
----------
- Fixed race condition on startup of tty and console threads.
- Fixed bug that leads to coredump when not selecting interface with tty.


2006/01/11
----------
- Fixed problem in xstp_com_other().
- Added if protocol.visible in some loops.
- Removed ARP protocol from yersinia.conf.


2005/12/30
----------
- Added default values on startup just to avoid sending packets with bad 
  values like MAC addresses filled with zeroes and so on...
- Fixed bad display showing the protocol params in the CLI.
- Initialize just the visible protocols on term_add_node().
- Fixed bug in ncurses protocol select screen ('g' key).
- Added all the functions keys to change protocol mode in ncurses interface.
- Removed nonvisible protocols with the ncurses function keys.
- Take into account a 0 bytes filename when pressing 'w' in ncurses.
- Removed the ncurses '!' option (not useful and broke the design).


2005/12/25
----------
- Added a new function to each protocol: protocol_end(struct term_node *).
  Called when destroying the node.


2005/12/22
----------
- Fixed bug in interfaces_get_last.
- Changed intefaces_add() to interfaces_enable().
- Changed intefaces_del() to interfaces_disable().
- Removed global mutex_int and created new mutex field for list_t.
- Normalized file names.
 

2005/12/21
----------
- Changed the interfaces list from static to a dynamic one. Now you can have
  as many interfaces as you wish!


2005/12/16
----------
- Ncurses split in three different files like GTK
- New files for dynamic lists implementation


2005/11/16
----------
- New CLI design!! My friend, we have TAB support...Yahoooo!! :)
- Fixed 3 annoying bugs on attack_kill_th :)


2005/11/08
----------
- Removed getopt from distribution (no longer needed).


2005/11/04
----------
- Added specific filter param for protocol in struct commands_param.
- Added xstp and cdp specific filter functions for params.


2005/11/03
----------
- First try for making an 'abstract' command line parser for any protocol.


2005/11/02
----------
- Moved attack_filter_param to parser_filter_param


2005/11/01
----------
- Fixed bad length on vtp CLI struct, dtp CLI struct, hsrp CLI struct and
  xstp CLI struct.
- Updated the maximum CLI command length from 32 to 48 bytes.
- Added keeping care of the 'active' protocols on interfaces engine (not

src/vtp.h
/* vtp.h
 * Definitions for Cisco's VLAN Trunking Protocol
 *
 * Yersinia
 * By David Barroso <tomac@yersinia.net> and Alfredo Andres <aandreswork@hotmail.com>
 * Copyright 2005-2017 Alfredo Andres and David Barroso
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
 
#ifndef __VTP_H__
#define __VTP_H__

#include <libnet.h>

#include "terminal-defs.h"
#include "interfaces.h"

#define VTP_SUMM_ADVERT   0x01
#define VTP_SUBSET_ADVERT 0x02
#define VTP_REQUEST       0x03
#define VTP_JOIN          0x04

static const struct tuple_type_desc vtp_code[] = {
    { VTP_SUMM_ADVERT,   "SUMMARY" },
    { VTP_SUBSET_ADVERT, "SUBSET"  },
    { VTP_REQUEST,       "REQUEST" },
    { VTP_JOIN,          "JOIN"    },
    { 0, NULL }
};


#define VLAN_TYPE_ETHERNET 0x01
#define VLAN_TYPE_FDDI     0x02
#define VLAN_TYPE_TRCRF    0x03
#define VLAN_TYPE_FDDI_NET 0x04
#define VLAN_TYPE_TRBRF    0x05

static const struct tuple_type_desc vlan_type[] = {
    { VLAN_TYPE_ETHERNET, "Ethernet" },
    { VLAN_TYPE_FDDI,     "FDDI"     },
    { VLAN_TYPE_TRCRF,    "TRCRF"    },
    { VLAN_TYPE_FDDI_NET, "FDDI-NET" },
    { VLAN_TYPE_TRBRF,    "TRBRF"    },
    { 0, NULL }
};



/* Default values */
#define VTP_DFL_VERSION 0x01
#define VTP_DFL_DOMAIN  "\x0\x0\x0\x0\x0\x0\x0\x0"
#define VTP_DFL_DOM_LEN 0x08
#define VTP_DFL_CODE    VTP_REQUEST

#define VTP_TIMESTAMP_SIZE 12
#define VTP_DOMAIN_SIZE    32

#define VLAN_MAX       64
#define VLAN_NAME_SIZE 32
#define VLAN_ALIGNED_LEN(x)   (4*(((x)+3)/4) )

#define VTP_DOT10_BASE 0x100000

#define VTP_VLAN_ADD     0x00
#define VTP_VLAN_DEL     0x01
#define VTP_VLAN_DEL_ALL 0x02

static struct proto_features vtp_features[] = {
    { F_LLC_CISCO, 0x2003 },
    { -1, 0 }
};




struct vlan_info_print  {
     u_int8_t  type;
     u_int16_t id;
     u_int32_t dot10;
     u_int8_t  name[VLAN_NAME_SIZE+1];
};

struct vlan_info {
     u_int8_t  len;   
     u_int8_t  status;
     u_int8_t  type;  
     u_int8_t  name_len;
     u_int16_t id;     
     u_int16_t mtu;    
     u_int32_t dot10;   
};

struct vtp_summary {
     u_int8_t  version;         
     u_int8_t  code;            
     u_int8_t  followers;       
     u_int8_t  dom_len;         
     u_int8_t  domain[VTP_DOMAIN_SIZE]; 
     u_int32_t revision;                    
     u_int32_t updater;                     
     u_int8_t  timestamp[VTP_TIMESTAMP_SIZE];    
     u_int8_t  md5[16];
};

struct vtp_subset {
     u_int8_t  version;     
     u_int8_t  code;        
     u_int8_t  seq;    
     u_int8_t  dom_len;
     u_int8_t  domain[VTP_DOMAIN_SIZE];
     u_int32_t revision;       
};

struct vtp_request {
     u_int8_t  version; 
     u_int8_t  code;    
     u_int8_t  reserved;
     u_int8_t  dom_len;
     u_int8_t  domain[VTP_DOMAIN_SIZE];
     u_int16_t start_val;   
};

struct vtp_join {
     u_int8_t  version;             
     u_int8_t  code;                
     u_int8_t  maybe_reserved;
     u_int8_t  dom_len;        
     u_int8_t  domain[VTP_DOMAIN_SIZE]; 
     u_int32_t vlan;          
     u_int8_t  unknown[126];
};


/* VTP mode stuff */
struct vtp_data { 
     u_int8_t  mac_source[ETHER_ADDR_LEN];
     u_int8_t  mac_dest[ETHER_ADDR_LEN];
     u_int8_t  version;
     u_int8_t  code;
     u_int8_t  followers;
     u_int8_t  seq;
     char      domain[VTP_DOMAIN_SIZE+1];
     u_int8_t  dom_len;
     u_int16_t start_val;
     u_int32_t revision;
     u_int32_t updater;
     u_int8_t  timestamp[VTP_TIMESTAMP_SIZE+1];
     u_int8_t  md5[16];
     u_int16_t vlans_len;
     u_int8_t  *vlan_info;
     u_int8_t  options[MAX_TLV*MAX_VALUE_LENGTH];
     u_int16_t options_len;
};


#define VTP_SMAC        0
#define VTP_DMAC        1
#define VTP_VERSION     2 
#define VTP_CODE        3
#define VTP_DOMAIN      4
#define VTP_MD5         5
#define VTP_UPDATER     6
#define VTP_REVISION    7
#define VTP_TIMESTAMP   8
#define VTP_STARTVAL    9
#define VTP_FOLLOWERS  10
#define VTP_SEQ        11
#define VTP_VLAN       14

/* Struct needed for using protocol fields within the network client */
struct commands_param vtp_comm_params[] = {
    { VTP_SMAC, "source",    "Source MAC", 6, FIELD_MAC, "Set source MAC address", 
                                        " H:H:H:H:H:H    48 bit mac address", 17, 1, 0, NULL, NULL },
    { VTP_DMAC, "dest",      "Destination MAC", 6, FIELD_MAC, "Set destination MAC address", 
                                        " H:H:H:H:H:H    48 bit mac address", 17, 1, 0, NULL, NULL },
    { VTP_VERSION, "version",   "Version", 1, FIELD_HEX, "Set vtp version", 
                                        " <00-FF>    virtual trunking version", 2, 2, 0, NULL, NULL },
    { VTP_CODE, "code",      "Code", 1, FIELD_HEX, "Set vtp code", 
                                        " <00-FF>    virtual trunking code", 2, 2, 1, NULL, vtp_code },
    { VTP_DOMAIN, "domain",    "Domain", VTP_DOMAIN_SIZE, FIELD_STR, "Set vtp domain name to use", 
                                        " WORD    Domain name", VTP_DOMAIN_SIZE, 2, 1, NULL, NULL },
    { VTP_MD5, "md5",       "MD5", 16, FIELD_BYTES, "Set vtp md5 hash", 
                                        " HHHHH...    MD5 hash", 32, 3, 1, NULL, NULL },
    { VTP_UPDATER, "updater",   "Updater", 4, FIELD_IP, "Set updater IP address", 
                                        " A.A.A.A    IPv4 address", 15, 3, 0, NULL, NULL },                
    { VTP_REVISION, "revision",  "Revision", 4, FIELD_DEC, "Set vtp revision number", 
                                        " <0-1999999999>    Revision number", 10, 4, 0, NULL, NULL },
    { VTP_TIMESTAMP, "timestamp", "Timestamp", VTP_TIMESTAMP_SIZE, FIELD_STR, "Set vtp timestamp", 
                                        " WORD    Timestamp text", VTP_TIMESTAMP_SIZE, 4, 0, NULL, NULL },
    { VTP_STARTVAL, "startval",  "Start value", 2, FIELD_DEC, "Set vtp start value", 
                                        " <0-65535>    Start value", 5, 4, 0, NULL, NULL },
    { VTP_FOLLOWERS, "followers", "Followers", 1,  FIELD_DEC, "Set vtp followers", 
                                        " <0-255>    Followers number", 3, 5, 0, NULL, NULL },
    { VTP_SEQ, "sequence",  "Sequence", 1, FIELD_DEC, "Set vtp sequence number", 
                                        " <0-255>    Sequence number", 3, 5, 0, NULL, NULL },
    { 0, "defaults",  NULL, 0, FIELD_DEFAULT, "Set all values to default", 
                                        " <cr>", 0, 0, 0, NULL, NULL }, 
    { 0, "interface", NULL, IFNAMSIZ, FIELD_IFACE, "Set network interface to use", 
                                        " WORD    Network interface", IFNAMSIZ, 0, 0, NULL, NULL },
    { VTP_VLAN, "vlan",     "VLAN", 0, FIELD_EXTRA, "", "", 0, 0, 0, NULL, NULL}
};


void vtp_th_send(void *);
void vtp_th_send_exit(struct attacks *);
void vtp_th_dos_del_all(void *);
void vtp_th_dos_del_all_exit(struct attacks *);
void vtp_th_dos_del(void *);
void vtp_th_dos_del_exit(struct attacks *);
void vtp_th_dos_add(void *);
void vtp_th_dos_add_exit(struct attacks *);
void vtp_th_dos_crash(void *);
void vtp_th_dos_crash_exit(struct attacks *);

#define VTP_PARAM_VLAN_ID    0
#define VTP_PARAM_VLAN_NAME  1

static struct attack_param vtp_vlan_add_param[] = {
    { NULL, "VLAN ID",   2,              FIELD_DEC, 4,              NULL },
    { NULL, "VLAN Name", VLAN_NAME_SIZE, FIELD_STR, VLAN_NAME_SIZE, NULL }
};

static struct attack_param vtp_vlan_del_param[] = {
    { NULL, "VLAN ID",   2, FIELD_DEC,  4, NULL }
};

#define VTP_ATTACK_SEND    0
#define VTP_ATTACK_DEL_ALL 1
#define VTP_ATTACK_DEL     2
#define VTP_ATTACK_ADD     3
#define VTP_ATTACK_CRASH   4


static struct _attack_definition vtp_attack[] = {
  { VTP_ATTACK_SEND,   "sending VTP packet",     NONDOS, SINGLE, vtp_th_send,        NULL,    0 },
  { VTP_ATTACK_DEL_ALL,"deleting all VTP vlans", DOS,    SINGLE, vtp_th_dos_del_all, NULL,    0 },
  { VTP_ATTACK_DEL,    "deleting one vlan",      DOS,    SINGLE, vtp_th_dos_del,     vtp_vlan_del_param,
                    SIZE_ARRAY(vtp_vlan_del_param) },
  { VTP_ATTACK_ADD,    "adding one vlan",        NONDOS, SINGLE, vtp_th_dos_add,     vtp_vlan_add_param,
                    SIZE_ARRAY(vtp_vlan_add_param) },
  { VTP_ATTACK_CRASH, "Catalyst zero day",       DOS,    SINGLE, vtp_th_dos_crash,   NULL,    0 },
  { 0,                  NULL,                    0,      0,      NULL,           NULL,    0 }
};



void    vtp_register(void);
int8_t  vtp_send(struct attacks *);
int8_t  vtp_init_attribs(struct term_node *);
int8_t  vtp_learn_packet(struct attacks *attacks, char *, u_int8_t *, void *, struct pcap_pkthdr *);
char    **vtp_get_printable_packet(struct pcap_data *);
char    **vtp_get_printable_store(struct term_node *);
int8_t  vtp_load_values(struct pcap_data *, void *);
int8_t  vtp_update_field(int8_t, struct term_node *, void *);
int8_t  vtp_generate_md5(char *, u_int32_t, u_int32_t, char *, u_int8_t, u_int8_t *, u_int16_t, u_int8_t *, u_int8_t);
int8_t  vtp_del_vlan(u_int16_t, u_int8_t *, u_int16_t *);
void    vtp_modify_vlan(u_int8_t, struct attacks *);
int8_t  vtp_add_vlan(u_int16_t, char *, u_int8_t **, u_int16_t *);
int8_t  vtp_init_comms_struct(struct term_node *);
int8_t  vtp_end(struct term_node *);

extern void   thread_libnet_error( char *, libnet_t *);
extern int8_t thread_create(pthread_t *, void *, void *);
extern void   write_log( u_int16_t mode, char *msg, ... );
extern int8_t attack_th_exit(struct attacks *);
extern void   attack_gen_mac(u_int8_t *);
extern struct interface_data *interfaces_get_packet(list_t *, struct interface_data *, u_int8_t *, struct pcap_pkthdr *, u_int8_t *, u_int16_t, time_t);
extern int8_t parser_vrfy_mac(char *, u_int8_t *);
extern int8_t parser_get_inet_aton(char *, struct in_addr *);
extern int8_t parser_get_formated_inet_address(u_int32_t, char *, u_int16_t);
extern void   md5_sum(const u_char *, size_t, u_char *);

extern int8_t parser_command2index(register const struct _attack_definition *, register int8_t);
extern struct terminals *terms;

extern int8_t bin_data[];

#endif

/* vim:set tabstop=4:set expandtab:set shiftwidth=4:set textwidth=120: */
